src/Corax/Queries/SuggestionTermProvider.cs

```diff
@@ -1,140 +1,481 @@
+                Slice.From(ctx, "PostingLists", ByteStringType.Immutable, out PostingListsSlice);
+                Slice.From(ctx, "Entries", ByteStringType.Immutable, out EntriesContainerSlice);
+                Slice.From(ctx, "NumberOfEntries", ByteStringType.Immutable, out NumberOfEntriesSlice);
+            }
+        }
+
+        public SuggestionTermProvider(
+            IndexSearcher searcher, int fieldId,
+            Slice term, IndexFieldBinding binding, int take, bool sortByPopularity, float accuracy, TDistanceProvider distanceProvider,
+            delegate*<ref SuggestionTermProvider<TDistanceProvider>, ref Span<byte>, ref Span<Token>, ref Span<float>, void> nextFunc)
+        {            
             _searcher = searcher;
             _fieldId = fieldId;
             _term = term;
-            _analyzer = analyzer;
+            _binding = binding;
             _take = take;
+            _distanceProvider = distanceProvider;
+            _distance = accuracy;
             _nextFunc = nextFunc;
+            _sort = sortByPopularity;
         }
 
-        private struct ScoreComparer : IComparer<(Slice, int)>
+        private struct ScoreComparer : IComparer<(Slice, float)>, IComparer<(Slice, int)>
         {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public int Compare((Slice, float) x, (Slice, float) y)
+            {
+                float result = y.Item2 - x.Item2;
+                return Math.Abs(result) < Constants.Boosting.ScoreEpsilon ? 0 : Math.Sign(result);
+            }
+
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
             public int Compare((Slice, int) x, (Slice, int) y)
             {
-                return x.Item2 - y.Item2;
+                return y.Item2 - x.Item2;
             }
         }
 
-        public static SuggestionTermProvider YieldFromNGram(IndexSearcher searcher, int fieldId, Slice term, Analyzer analyzer, int take = -1)
+        private struct SuggestionsNGramTable
         {
-            static void Next(ref SuggestionTermProvider provider, ref Span<byte> terms, ref Span<Token> tokens)
+            private int _count;
+            private int _currentIdx;
+
+            private byte[] _storage;
+            private int[] _gramPosition;
+            private int[] _gramSize;
+            private float[] _gramBoost;
+
+            public readonly int MaxGramSize;
+            public int Count => _count;
+
+            public SuggestionsNGramTable(int maxGramSize)
+            {
+                MaxGramSize = maxGramSize;
+
+                _count = 0;
+                _currentIdx = -1;
+
+                _storage = null;
+                _gramPosition = null;
+                _gramSize = null;
+                _gramBoost = null;
+            }
+
+            public void Generate(ReadOnlySpan<byte> word)
+            {
+                // Copy to storage.
+                _storage = new byte[word.Length];
```

Comment:
Isn't that something to avoid?

Reply:
Same as the ones below? This is being used for many words, no? Can't we reuse the buffers, at least?

Reply:
Alternatively, `ArrayPool`, if this isn't being reused?